src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,372 @@
+        }
+        if new.perm == Permission::Unique {
+            // Make sure the possibly-unique range contains the new borrow
+            self.unique_range.start = self.unique_range.start.min(new_idx);
+            self.unique_range.end = self.unique_range.end.max(new_idx + 1);
+        }
+
+        // The above insert changes the meaning of every index in the cache >= new_idx, so now
+        // we need to find every one of those indexes and increment it.
+        // But if the insert is at the end (equivalent to a push), we can skip this step because
+        // it didn't change the position of any other tags.
+        if new_idx != self.borrows.len() - 1 {
+            for idx in &mut self.cache.idx {
+                if *idx >= new_idx {
+                    *idx += 1;
+                }
+            }
+        }
+
+        // This primes the cache for the next access, which is almost always the just-added tag.
+        self.cache.add(new_idx, new.tag);
+
+        #[cfg(feature = "expensive-debug-assertions")]
+        self.verify_cache_consistency();
+    }
+
+    /// Construct a new `Stack` using the passed `Item` as the base tag.
+    pub fn new(item: Item) -> Self {
+        Stack {
+            borrows: vec![item],
+            unknown_bottom: None,
+            #[cfg(feature = "stack-cache")]
+            cache: StackCache { idx: [0; CACHE_LEN], tags: [item.tag; CACHE_LEN] },
+            #[cfg(feature = "stack-cache")]
+            unique_range: if item.perm == Permission::Unique { 0..1 } else { 0..0 },
+        }
+    }
+
+    pub fn get(&self, idx: usize) -> Option<Item> {
+        self.borrows.get(idx).cloned()
+    }
+
+    #[allow(clippy::len_without_is_empty)] // Stacks are never empty
+    pub fn len(&self) -> usize {
+        self.borrows.len()
+    }
+
+    pub fn unknown_bottom(&self) -> Option<SbTag> {
+        self.unknown_bottom
+    }
+
+    pub fn set_unknown_bottom(&mut self, tag: SbTag) {
+        // We clear the borrow stack but the lookup cache doesn't support clearing per se. Instead,
+        // there is a check explained in `find_granting_cache` which protects against accessing the
+        // cache when it has been cleared and not yet refilled.
+        self.borrows.clear();
+        self.unknown_bottom = Some(tag);
+    }
+
+    /// Find all `Unique` elements in this borrow stack above `granting_idx`, pass a copy of them
+    /// to the `visitor`, then set their `Permission` to `Disabled`.
+    pub fn disable_uniques_starting_at<V: FnMut(Item) -> crate::InterpResult<'tcx>>(
+        &mut self,
+        disable_start: usize,
+        mut visitor: V,
+    ) -> crate::InterpResult<'tcx> {
+        #[cfg(feature = "stack-cache")]
+        let unique_range = self.unique_range.clone();
+        #[cfg(not(feature = "stack-cache"))]
+        let unique_range = 0..self.len();
+
+        if disable_start <= unique_range.end {
+            // add 1 so we don't disable the granting item
```

Comment:
I don't understand this comment.

Reply:
Because it's leftover from when the argument to that function was the granting index. Easy fix.